okta/cache/cache.go

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2018 - Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cache
+
+import (
+	"net/http"
+)
+
+type Cache interface {
+	Get(key string) *http.Response
+	Set(key string, value *http.Response)
+	Delete(key string)
+	Clear()
+	Has(key string) bool
+}
+
+func CreateCacheKey(req *http.Request) string {
+	s := req.URL.Scheme + "://" + req.URL.Host + req.URL.RequestURI()
```

Comment:
this is actually a pretty complex thing here, I understand it's purpose is local caching, but consider this vulnerability that just came out against Cloudflare and other CDNs:

https://blog.cloudflare.com/cache-poisoning-protection/

https://portswigger.net/blog/practical-web-cache-poisoning

It tends to make me think that in the future considering other HTTP headers, like even headers for Conditional Gets (eg, If-Modified-Since) will make sense?  Have the SDKs in other languages dealt with this?

Reply:
Is this a concern for the SDKs?  These libraries build requests based on either the configured okta domain, or by responses from the Okta API.  So I'm not sure where the user-supplied url content would come from?

Reply:
The SDK's as far as i know don't do anything with these headers.

Reply:
Specifically that as a class of problems, it is difficult to accurately represent an HTTP Request with just the scheme, host and Request URI.

For example, if a request contained an `If-None-Match` header, a client would expect a 304 not modified with an empty body, or a new response.   But a later request could get the cached 304, even if it didn't include the `If-None-Match`.  Basically there is a mismatch between the cache key, and reality of how a server interprets the request.

This is more a here-be-dragons warning, that HTTP Request canonicalization is a subject that even CDNs and people doing HTTP all the time can get wrong -- and this specific canonicalization can have issues with common use cases like conditional HTTP GETs (which after talking to @bretterer in person and Karl, it sounds like no Okta APIs support E-Tags)

Reply:
I currently check for a `GET` request before allowing the cache item to be set, would it be worth while to also check to see if it is in the 200-299 status code range here as well?